packages/viewer/src/scenes/operations.ts

```diff
@@ -0,0 +1,241 @@
+      throw new Error('Cannot build selector. A selector has not been defined');
+    }
+    return this.query;
+  }
+
+  public or(): SelectorBuilder<OrSelector> {
+    return new OrSelectorBuilder(this);
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return new AndSelectorBuilder(this);
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.query = { type: 'supplied-id', value: suppliedId };
+    return this;
+  }
+
+  public withItemId(itemId: string): this {
+    this.query = { type: 'item-id', value: itemId };
+    return this;
+  }
+}
+
+/**
+ * A selector builder to perform boolean `or` operations.
+ */
+export class OrSelectorBuilder implements SelectorBuilder<OrSelector> {
+  private builders?: SelectorBuilder<ItemSelector>[] = [];
+
+  private parent: SelectorBuilder<ItemSelector>;
+  public constructor(parent: SelectorBuilder<ItemSelector>) {
+    this.builders.push(parent);
+    this.parent = parent;
+  }
+
+  public or(): SelectorBuilder<OrSelector> {
+    return this;
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return new AndSelectorBuilder(this.parent);
+  }
+
+  public build(): OrSelector {
+    return { type: 'or', selectors: this.builders.map(q => q.build()) };
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withSuppliedId(suppliedId));
+    return this;
+  }
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * A selector builder to perform boolean `and` operations.
+ */
+export class AndSelectorBuilder implements SelectorBuilder<AndSelector> {
+  private builders?: SelectorBuilder<ItemSelector>[] = [];
+  private parent: SelectorBuilder<ItemSelector>;
+  public constructor(parent: SelectorBuilder<ItemSelector>) {
+    this.builders.push(parent);
+    this.parent = parent;
+  }
+
+  public or(): SelectorBuilder<OrSelector> {
+    return new OrSelectorBuilder(this.parent);
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return this;
+  }
+
+  public build(): AndSelector {
+    return { type: 'and', selectors: this.builders.map(q => q.build()) };
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withSuppliedId(suppliedId));
+    return this;
+  }
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * An `ItemSelector` describes a type for performing a selection against some
+ * attribute of a scene item.
+ */
+export type ItemSelector = SuppliedIdSelector | ItemIdSelector;
+
+interface OrSelector {
+  type: 'or';
+  selectors: ItemSelector[];
+}
+
+interface AndSelector {
+  type: 'and';
+  selectors: ItemSelector[];
+}
+
+export interface OperationDefinition {
+  operation: ItemOperation | SceneOperation;
+}
+
+export interface SelectorBuilder<T> {
```

Comment:
We also need to handle queries that operate on everything, e.g. `showAll()`, `hideAll()`, `clearAllMaterialOverrides()`. Maybe we could so something like:

```ts
scene.where(q => q.all())
  .show()
  .execute()
```

When calling `.all()` it should return a terminal type that doesn't allow further chaining.

Reply:
Is there backend support for this yet?

Reply:
Chatted with Jared/Kevin about this. Currently there is no backend support for the showAll, hideAll, clear all. Might be worth doing as a follow up item.